{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNn4VKuDHW91+SAVFvDry6O",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/JesseK18/NaCo1/blob/Jesse/NACO1colab.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install \"geopandas<1.0.0\" numpy matplotlib pandas svgpath2mpl"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "H7J7ltsbIp4l",
        "outputId": "0b200195-2060-4817-bdc4-b7f875121ceb",
        "collapsed": true
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting geopandas<1.0.0\n",
            "  Downloading geopandas-0.14.4-py3-none-any.whl.metadata (1.5 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (1.26.4)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.10/dist-packages (3.8.0)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (2.2.2)\n",
            "Collecting svgpath2mpl\n",
            "  Downloading svgpath2mpl-1.0.0-py2.py3-none-any.whl.metadata (2.3 kB)\n",
            "Collecting fiona>=1.8.21 (from geopandas<1.0.0)\n",
            "  Downloading fiona-1.10.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (56 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.6/56.6 kB\u001b[0m \u001b[31m1.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from geopandas<1.0.0) (24.1)\n",
            "Requirement already satisfied: pyproj>=3.3.0 in /usr/local/lib/python3.10/dist-packages (from geopandas<1.0.0) (3.7.0)\n",
            "Requirement already satisfied: shapely>=1.8.0 in /usr/local/lib/python3.10/dist-packages (from geopandas<1.0.0) (2.0.6)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.3.0)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (4.54.1)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.4.7)\n",
            "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (10.4.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (3.2.0)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas) (2024.2)\n",
            "Requirement already satisfied: attrs>=19.2.0 in /usr/local/lib/python3.10/dist-packages (from fiona>=1.8.21->geopandas<1.0.0) (24.2.0)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.10/dist-packages (from fiona>=1.8.21->geopandas<1.0.0) (2024.8.30)\n",
            "Requirement already satisfied: click~=8.0 in /usr/local/lib/python3.10/dist-packages (from fiona>=1.8.21->geopandas<1.0.0) (8.1.7)\n",
            "Collecting click-plugins>=1.0 (from fiona>=1.8.21->geopandas<1.0.0)\n",
            "  Downloading click_plugins-1.1.1-py2.py3-none-any.whl.metadata (6.4 kB)\n",
            "Collecting cligj>=0.5 (from fiona>=1.8.21->geopandas<1.0.0)\n",
            "  Downloading cligj-0.7.2-py3-none-any.whl.metadata (5.0 kB)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.7->matplotlib) (1.16.0)\n",
            "Downloading geopandas-0.14.4-py3-none-any.whl (1.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.1/1.1 MB\u001b[0m \u001b[31m14.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading svgpath2mpl-1.0.0-py2.py3-none-any.whl (7.8 kB)\n",
            "Downloading fiona-1.10.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m17.3/17.3 MB\u001b[0m \u001b[31m21.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading click_plugins-1.1.1-py2.py3-none-any.whl (7.5 kB)\n",
            "Downloading cligj-0.7.2-py3-none-any.whl (7.1 kB)\n",
            "Installing collected packages: cligj, click-plugins, fiona, svgpath2mpl, geopandas\n",
            "  Attempting uninstall: geopandas\n",
            "    Found existing installation: geopandas 1.0.1\n",
            "    Uninstalling geopandas-1.0.1:\n",
            "      Successfully uninstalled geopandas-1.0.1\n",
            "Successfully installed click-plugins-1.1.1 cligj-0.7.2 fiona-1.10.1 geopandas-0.14.4 svgpath2mpl-1.0.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BO5x7htEIceQ",
        "outputId": "a18df19c-85c9-4108-d0f9-28e63387eb2e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "27107.43269686869\n",
            "27714.56363632267\n",
            "27439.128446492366\n",
            "27078.21447233571\n",
            "28323.600409591767\n",
            "26809.900755508486\n",
            "26946.838038116624\n",
            "26489.237770193355\n",
            "27802.937142168863\n",
            "27471.189937186573\n",
            "27077.24790232253\n",
            "27154.73992447906\n",
            "27373.292807893795\n",
            "28659.379703529183\n",
            "26893.753187136273\n",
            "28541.75428259978\n",
            "26039.19881548429\n",
            "27782.082621300444\n",
            "28585.20903774119\n",
            "27120.751200592025\n",
            "30018.57129598227\n",
            "28369.39402996109\n",
            "27040.854584946974\n",
            "26323.679965473977\n",
            "26713.25590326716\n",
            "27771.91130059903\n",
            "29716.021183677054\n",
            "28849.09169189078\n",
            "27821.11313730734\n",
            "26979.074435823208\n",
            "28608.365027473686\n",
            "26644.103203537048\n",
            "27458.596041421446\n",
            "28807.53486854108\n",
            "27038.736134996314\n",
            "27201.28550559206\n",
            "27680.435341821798\n",
            "28096.58398130367\n",
            "25980.056883750116\n",
            "26253.382780893862\n",
            "26507.34075226624\n",
            "27947.527762931113\n",
            "27067.756692812625\n",
            "28486.30089730517\n",
            "27284.962812043635\n",
            "27393.293491535846\n",
            "27778.762976340233\n",
            "29345.051609840917\n",
            "27485.902581679835\n",
            "26629.678929374746\n",
            "27085.455568054917\n"
          ]
        }
      ],
      "source": [
        "\"\"\"README\n",
        "\n",
        "This file contains all the code needed the assignment 1 for the Natural Computing\n",
        "course 24/25. Please see the example at the bottom of this file on how to interact\n",
        "with the TSP object.\n",
        "\n",
        "Installation\n",
        "------------\n",
        "    Call this in a terminal:\n",
        "        pip install \"geopandas<1.0.0\" numpy matplotlib pandas svgpath2mpl\n",
        "\n",
        "\"\"\"\n",
        "\n",
        "import warnings\n",
        "import math\n",
        "import io\n",
        "import typing\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib as mpl\n",
        "import matplotlib.pyplot as plt\n",
        "import svgpath2mpl\n",
        "import geopandas\n",
        "\n",
        "\n",
        "DATA = \"\"\"hckey,capital,capital_lat,capital_lng\n",
        "ad,Andorra,42.5,1.5165\n",
        "al,Tirana,41.3275,19.8189\n",
        "at,Vienna,48.2,16.3666\n",
        "ba,Sarajevo,43.85,18.383\n",
        "be,Brussels,50.8333,4.3333\n",
        "bg,Sofia,42.6833,23.3167\n",
        "by,Minsk,53.9,27.5666\n",
        "ch,Bern,46.9167,7.467\n",
        "cy,Nicosia,35.1667,33.3666\n",
        "cz,Prague,50.0833,14.466\n",
        "de,Berlin,52.5218,13.4015\n",
        "dk,Copenhagen,55.6786,12.5635\n",
        "ee,Tallinn,59.4339,24.728\n",
        "es,Madrid,40.4,-3.6834\n",
        "fi,Helsinki,60.1756,24.9341\n",
        "fr,Paris,48.8667,2.3333\n",
        "gb,London,51.5072,-0.1275\n",
        "gr,Athens,37.9833,23.7333\n",
        "hr,Zagreb,45.8,16\n",
        "hu,Budapest,47.5,19.0833\n",
        "ie,Dublin,53.3331,-6.2489\n",
        "is,Reykjavik,64.15,-21.95\n",
        "it,Rome,41.896,12.4833\n",
        "li,Vaduz,47.1337,9.5167\n",
        "lt,Vilnius,54.6834,25.3166\n",
        "lu,Luxembourg,49.6117,6.13\n",
        "lv,Riga,56.95,24.1\n",
        "mc,Monaco,43.7396,7.4069\n",
        "md,Chisinau,47.005,28.8577\n",
        "me,Podgorica,42.466,19.2663\n",
        "mk,Skopje,42,21.4335\n",
        "mt,Valletta,35.8997,14.5147\n",
        "nl,Amsterdam,52.35,4.9166\n",
        "no,Oslo,59.9167,10.75\n",
        "pl,Warsaw,52.25,21\n",
        "pt,Lisbon,38.7227,-9.1449\n",
        "ro,Bucharest,44.4334,26.0999\n",
        "rs,Belgrade,44.8186,20.468\n",
        "ru,Moscow,55.7522,37.6155\n",
        "se,Stockholm,59.3508,18.0973\n",
        "si,Ljubljana,46.0553,14.515\n",
        "sk,Bratislava,48.15,17.117\n",
        "sm,San Marino,43.9172,12.4667\n",
        "ua,Kiev,50.4334,30.5166\"\"\"\n",
        "\n",
        "LEIDEN_SVG = \"\"\"M380.52 239.51l-30.679-29.744-29.757-34.948-14.23-13.797v-13.258l-33.266-32.255-10.352 10.033-7.763-7.524-10.348 10.034 5.917 5.734-17.746\n",
        "17.203-14.047-13.62 10.721-10.395-10.351-10.033-10.718 10.395-12.197-11.827 10.717-10.395-11.452-11.107-10.721 10.391-16.266-15.767\n",
        "2.957-2.868 18.481-17.918 11.46 11.107 9.982-9.675-6.656-6.45 17.004-16.488-6.653-6.45 14.416-13.978 120.868 117.197\n",
        "29.754 31.71v18.457L403.9 216.84h20.608c2.352 0 8.24-8.424 10.436-10.142 6.777-5.306 13.24-4.346 21.31-3.43 13.316 1.514 23.955\n",
        "9.485 33.313 18.203 26.695 24.87 21.382 53.31.735 79.32.788-.988 7.349 10.174 7.555 11.301.79 4.372-.726 9.239-3.29 12.907-5.464\n",
        "7.82-15.208 8.218-24.226 8.231l-6.744-6.542c-11.747 11.39-16.922 16.034-33.659 16.037-31.782.003-53.845-10.81-66.787-40.495v-25.804l4.434-12.904\n",
        "12.936-12.543V239.51zm65.036 69.75c27.483-15.113 49.34-48.926 26.594-70.203-11.619-10.87-22.464-10.176-36.766-4.208-.032 0 .252\n",
        "30.598-4.734 30.598l-5.029 4.879-29.907.402c-14.831 14.377-7.986 36.849 11.629 44.275 11.318 4.286 25.378 1.311 38.213-5.744zm55.91\n",
        "13.359l-5.166-10.18 5.166 10.18zm-298.33-6.814l-3.457 6.814 3.458-6.814a14.054 14.054 0 0 0 1.152-3.714c.206-1.127\n",
        "6.767-12.289 7.555-11.3-20.647-26.01-25.96-54.45.735-79.32 9.358-8.72 19.997-16.69 33.312-18.205 8.07-.915 14.534-1.875\n",
        "21.31 3.431 2.197 1.718 8.085 10.142 10.437 10.142h20.608l24.304-23.565v-18.457l29.754-31.71L472.172 25.91l14.416\n",
        "13.978-6.653 6.45 17.004 16.487-6.656 6.45 9.983 9.676 11.459-11.107 18.48 17.918 2.958 2.868-16.266 15.767-10.72-10.391-11.453\n",
        "11.107 10.717 10.395-12.197 11.827-10.718-10.395-10.351 10.033 10.72 10.395-14.046 13.62-17.746-17.203 5.917-5.734-10.348-10.034-7.763\n",
        "7.524-10.352-10.033-33.266 32.255v13.258l-14.23 13.797-29.757 34.948-30.679 29.744v11.468l12.936 12.543 4.434 12.904v25.804c-12.942\n",
        "29.685-35.004 40.498-66.787 40.495-16.737-.003-21.912-4.648-33.659-16.037l-6.744 6.542c-9.018-.013-18.762-.412-24.225-8.23-1.854-2.652-3.16-5.93-3.443-9.194zm293.125-3.444L490.491\n",
        "301l5.77 11.36zm-240.672-3.102c12.835 7.055 26.895 10.03 38.213 5.744 19.615-7.426 26.46-29.898 11.63-44.275l-29.908-.402-5.029-4.879c-4.986\n",
        "0-4.702-30.598-4.734-30.598-14.302-5.968-25.147-6.662-36.766 4.208-22.745 21.277-.889 55.09 26.594 70.202z\n",
        "\"\"\"\n",
        "LEIDEN = 4.497010, 52.160114\n",
        "\n",
        "\n",
        "def plot_europe(\n",
        "    data: pd.DataFrame, fig: plt.Figure = None, ax: plt.Axes = None\n",
        ") -> typing.Tuple[plt.Figure, plt.Axes]:\n",
        "    \"\"\"Plotting utilitly, plots a map of Europe, with Leiden explitly marked\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    data: pd.DataFrame\n",
        "        Locations of cities on the map\n",
        "\n",
        "    fig: mpl.Figure (optional)\n",
        "    ax: plt.Axes (optional)\n",
        "        Optional figure and axes for plotting\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "        (mpl.Figure, plt.Axes,)\n",
        "            Handles to the plot\n",
        "    \"\"\"\n",
        "\n",
        "    if fig is None:\n",
        "        fig, ax = plt.subplots(1, 1, figsize=(15, 8))\n",
        "\n",
        "    keys = svgpath2mpl.parse_path(LEIDEN_SVG)\n",
        "    keys.vertices -= keys.vertices.mean(axis=0)\n",
        "\n",
        "    keys = keys.transformed(mpl.transforms.Affine2D().rotate_deg(180))\n",
        "\n",
        "    with warnings.catch_warnings():\n",
        "        warnings.simplefilter(\"ignore\")\n",
        "        world = geopandas.read_file(geopandas.datasets.get_path(\"naturalearth_lowres\"))\n",
        "\n",
        "    world.plot(ax=ax, color=\"lightgray\", edgecolor=\"black\", alpha=0.5)\n",
        "\n",
        "    ax.scatter(data[\"capital_lng\"], data[\"capital_lat\"], s=10, color=\"blue\", alpha=0.5)\n",
        "\n",
        "    ax.scatter(*LEIDEN, color=\"red\", marker=keys, s=5_000, alpha=0.6)\n",
        "    ax.scatter(*LEIDEN, color=\"red\")\n",
        "    ax.text(*LEIDEN, \"Leiden\", ha=\"right\", **{\"fontsize\": 11})\n",
        "\n",
        "    for _, c in data.iterrows():\n",
        "        ax.text(c[\"capital_lng\"], c[\"capital_lat\"], c[\"capital\"], **{\"fontsize\": 10})\n",
        "\n",
        "    ax.set_xlim(-24, 40)\n",
        "    ax.set_ylim(34, 70)\n",
        "    plt.tight_layout()\n",
        "    return fig, ax\n",
        "\n",
        "\n",
        "def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:\n",
        "    \"\"\"\n",
        "    Calculate the great circle distance in kilometers between two points\n",
        "    on the earth (specified in decimal degrees)\n",
        "\n",
        "    Parameters\n",
        "    ----------\n",
        "    lat1: float\n",
        "        Latitude for point 1\n",
        "    lon1: float\n",
        "        Longtitude for point 1\n",
        "    lat2: float\n",
        "        Latitude for point 2\n",
        "    lon2: float\n",
        "        Longtitude for point 1\n",
        "\n",
        "    Returns\n",
        "    -------\n",
        "        float\n",
        "            The Haversine distance between point1 and point 2\n",
        "    \"\"\"\n",
        "\n",
        "    lon1, lat1, lon2, lat2 = map(math.radians, [lon1, lat1, lon2, lat2])\n",
        "\n",
        "    dlon = lon2 - lon1\n",
        "    dlat = lat2 - lat1\n",
        "    a = (\n",
        "        math.sin(dlat / 2) ** 2\n",
        "        + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2\n",
        "    )\n",
        "    c = 2 * math.asin(math.sqrt(a))\n",
        "    r = 6371\n",
        "    return c * r\n",
        "\n",
        "\n",
        "class TSP:\n",
        "    \"\"\"Traveling Salesperson object, with plotting utility\"\"\"\n",
        "\n",
        "    def __init__(self, plot: bool = True): # turn True for visualisation\n",
        "        \"\"\"Create a Traveling Salesperson object\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        plot: bool = True\n",
        "            Whether to create an (interactive) plot. When running this for the optimization\n",
        "            its advised to turn this off, as it can take quite a bit of extra time to visualize\n",
        "            the tour on every function call.\n",
        "        \"\"\"\n",
        "\n",
        "        self.data = pd.read_csv(io.StringIO(DATA))\n",
        "        self.plot = plot\n",
        "        self.line = None\n",
        "        self.dim = len(self.data)\n",
        "\n",
        "    def __enter__(self):\n",
        "        \"\"\"Create a plot, i.e. figure and axes, if self.plot == True.\"\"\"\n",
        "\n",
        "        if self.plot:\n",
        "            plt.ion()\n",
        "            _, self.ax = plot_europe(self.data)\n",
        "        return self\n",
        "\n",
        "    def __exit__(self, *args, **kwargs):\n",
        "        \"\"\"Stop plotting interactively, but keep showing the plot if self.plot == True.\"\"\"\n",
        "\n",
        "        if self.plot:\n",
        "            plt.ioff()\n",
        "            plt.show()\n",
        "\n",
        "    def __call__(self, path_idx: np.array) -> float:\n",
        "        \"\"\"Calculate the route length of the TSP problem.\n",
        "        The method returns the total length of the tour, as given by the order specified\n",
        "        by path_idx, caculated as the sucessive Haversine distance between cities.\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        path_idx: np.ndarray[int]\n",
        "            Indices of the cities you will visit, i.e. the order. This should be an integer\n",
        "            array of size n, where n is the number of cities. Every integer from 1 to n\n",
        "            should appear in this array, in arbitrary order, but no duplicates are allowed.\n",
        "\n",
        "        Returns\n",
        "        -------\n",
        "            float The length of the tour\n",
        "        \"\"\"\n",
        "        assert len(path_idx) == len(self.data), \"Make sure you visit all cities\"\n",
        "        assert len(set(path_idx)) == len(path_idx), \"Make sure all cities are unique\"\n",
        "\n",
        "        path = self.create_path(path_idx)\n",
        "\n",
        "        current = path[0]\n",
        "        route_length = 0\n",
        "\n",
        "        for stop in path[1:]:\n",
        "            route_length += haversine(*current, *stop)\n",
        "            current = stop\n",
        "\n",
        "        return route_length\n",
        "\n",
        "    def create_path(self, path_idx: np.ndarray) -> np.ndarray:\n",
        "        \"\"\"Convert an integer path to a matrix of lng, lat values, with Leiden pre- and appended\"\"\"\n",
        "\n",
        "        return np.vstack(\n",
        "            [\n",
        "                LEIDEN,\n",
        "                self.data.iloc[path_idx][[\"capital_lng\", \"capital_lat\"]].values,\n",
        "                LEIDEN,\n",
        "            ]\n",
        "        )\n",
        "\n",
        "    def plot_route(self, path: np.ndarray, route_length: float = float(\"inf\")) -> None:\n",
        "        \"\"\"Plot the route on the map of Europe, interactively.\n",
        "\n",
        "        Parameters\n",
        "        ----------\n",
        "        path: np.ndarray\n",
        "            The path of the TSP, can be either a list of indices, or a matrix\n",
        "            of lng, lat values.\n",
        "        route_length: float = inf (optional)\n",
        "            The length of the route, to display in the plot, optional.\n",
        "\n",
        "        \"\"\"\n",
        "        if self.plot:\n",
        "            if len(path.shape) == 1:\n",
        "                path = self.create_path(path)\n",
        "\n",
        "            if self.line is not None:\n",
        "                self.line[0].remove()\n",
        "                self.legend.remove()\n",
        "\n",
        "            self.line = self.ax.plot(\n",
        "                path[:, 0],\n",
        "                path[:, 1],\n",
        "                color=\"green\",\n",
        "                linestyle=\"--\",\n",
        "                alpha=0.9,\n",
        "                label=f\"route length: {route_length:.2f} km\",\n",
        "            )\n",
        "            self.legend = plt.legend()\n",
        "            plt.draw()\n",
        "            plt.pause(0.0001)\n",
        "\n",
        "class GA:\n",
        "    \"\"\"Genetic Algorithm experiment and plot\"\"\"\n",
        "\n",
        "    def __init__(self, tsp: TSP, POPULATION_SIZE=100, base_mutation_rate=0.1, increase_rate=0.05, decrease_rate=0.01):\n",
        "        self.POPULATION_SIZE = POPULATION_SIZE\n",
        "        self.POPULATION = np.array([np.random.permutation(tsp.dim) for _ in range(self.POPULATION_SIZE)])\n",
        "        self.tsp = tsp\n",
        "        self.base_mutation_rate = base_mutation_rate\n",
        "        self.mutation_rate = base_mutation_rate\n",
        "        self.increase_rate = increase_rate\n",
        "        self.decrease_rate = decrease_rate\n",
        "        self.no_improvement_count = 0\n",
        "        self.best_fitness = float('inf')\n",
        "        self.MUTATION_TYPE = ['insert_mutation', 'swap_mutation', 'inversion_mutation', 'scramble_mutation']\n",
        "\n",
        "    # Evaluating all routes, lengths of all total path lengths\n",
        "    def evaluate(self, population):\n",
        "        fitness_scores = np.array([tsp(chromosome) for chromosome in population])\n",
        "        return fitness_scores\n",
        "\n",
        "    # Selection by direct comparison of two routes, only shorter route selected to parents\n",
        "    def selection(self, fitness_scores, population, Q=2, REPETITIONS=50):\n",
        "        parents = []\n",
        "        for n in range(REPETITIONS):\n",
        "            tournament_selection_idx = np.random.choice(np.arange(len(population)), Q, replace=False) # maybe replace = True\n",
        "            tournament_winner = tournament_selection_idx[np.argmin(fitness_scores[tournament_selection_idx])]\n",
        "            parents.append(population[tournament_winner])\n",
        "        return parents\n",
        "\n",
        "    # Combines all parents randomly to create children according to cycle_cross_over\n",
        "    def cross_over(self, parents, C_O_RATE=.7):\n",
        "        offspring = []\n",
        "        n_iterations = len(parents)\n",
        "        for _ in range(n_iterations//2):\n",
        "            parent1_idx, parent2_idx = np.random.choice(n_iterations, 2, replace=False)\n",
        "            parent1 = parents[parent1_idx]\n",
        "            parent2 = parents[parent2_idx]\n",
        "            child1, child2 = self.cycle_cross_over(parent1, parent2)\n",
        "            offspring.append(child1)\n",
        "            offspring.append(child2)\n",
        "        return offspring\n",
        "\n",
        "    # Creates children from parents while maintaining completeness of genes\n",
        "    def cycle_cross_over(self, parent1, parent2):\n",
        "        size = len(parent1)\n",
        "        child1 = np.full(size, -1)\n",
        "        child2 = np.full(size, -1)\n",
        "        visited = np.zeros(size, dtype=bool)\n",
        "        gen_p1 = parent1[0]\n",
        "        stop_gen = parent1[0]\n",
        "        cycle = [0]\n",
        "        while True:\n",
        "            gen_p2_idx = np.where(parent2 == gen_p1)[0][0]\n",
        "            cycle.append(gen_p2_idx)\n",
        "            gen_p1 = parent1[gen_p2_idx]\n",
        "            if gen_p1 == stop_gen:\n",
        "                break\n",
        "        child1[cycle] = parent1[cycle]\n",
        "        child2[cycle] = parent2[cycle]\n",
        "        child1[child1 < 0] = parent2[child1 < 0]\n",
        "        child2[child2 < 0] = parent1[child2 < 0]\n",
        "        return child1, child2\n",
        "\n",
        "    # Mutation type setting (hybrid)\n",
        "    def mutations(self, population, mutation_rate=.1, mutation='all'): #or specific mutation\n",
        "        mutated_population = []\n",
        "        copy_population = np.copy(population)\n",
        "        num_mutations = int(len(population) * mutation_rate)\n",
        "        mutation_indices = np.random.choice(copy_population.shape[0], num_mutations, replace=False)\n",
        "        mutation_pool = copy_population[mutation_indices]\n",
        "        mutation_probabilities = [0.3, 0.1, 0.5, 0.1]\n",
        "        if mutation == 'all':\n",
        "          for chromosome in mutation_pool:\n",
        "              mutation = np.random.choice(self.MUTATION_TYPE, p=mutation_probabilities)\n",
        "              if mutation == 'insert_mutation':\n",
        "                  mutant = self.insert_mutation(chromosome)\n",
        "              elif mutation == 'swap_mutation':\n",
        "                  mutant = self.swap_mutation(chromosome)\n",
        "              elif mutation == 'scramble_mutation':\n",
        "                  mutant = self.swap_mutation(chromosome)\n",
        "              else:\n",
        "                  mutant = self.inversion_mutation(chromosome)\n",
        "              mutated_population.append(mutant)\n",
        "        else:\n",
        "          if mutation == 'insert_mutation':\n",
        "              for chromosome in mutation_pool:\n",
        "                mutant = self.insert_mutation(chromosome)\n",
        "                mutated_population.append(mutant)\n",
        "          elif mutation == 'swap_mutation':\n",
        "              for chromosome in mutation_pool:\n",
        "                mutant = self.swap_mutation(chromosome)\n",
        "                mutated_population.append(mutant)\n",
        "          elif mutation == 'scramble_mutation':\n",
        "              for chromosome in mutation_pool:\n",
        "                mutant = self.swap_mutation(chromosome)\n",
        "                mutated_population.append(mutant)\n",
        "          else:\n",
        "              for chromosome in mutation_pool:\n",
        "                mutant = self.inversion_mutation(chromosome)\n",
        "                mutated_population.append(mutant)\n",
        "        return mutated_population\n",
        "\n",
        "    # Mutation hybrid adaptive\n",
        "    def adaptive_mutation(self, population):\n",
        "        # Track best route\n",
        "        current_best_fitness = np.min(self.evaluate(population))\n",
        "        if current_best_fitness < self.best_fitness:\n",
        "            # Improvement update\n",
        "            self.no_improvement_count = 0\n",
        "            self.best_fitness = current_best_fitness\n",
        "            # So, mutation rate should be slowed down slightly bcs correct momentum\n",
        "            self.mutation_rate = max(self.base_mutation_rate, self.mutation_rate - self.decrease_rate)\n",
        "        else:\n",
        "            # No improvement update\n",
        "            self.no_improvement_count += 1\n",
        "            if self.no_improvement_count >= 5: # vary for slower\\faster increase\n",
        "                self.mutation_rate = min(1.0, self.mutation_rate + self.increase_rate)\n",
        "        mutated_population = []\n",
        "        copy_population = np.copy(population)\n",
        "        num_mutations = int(len(population) * self.mutation_rate)\n",
        "        population = np.array(population) if isinstance(population, list) else population\n",
        "        num_mutations = min(max(1, num_mutations), len(population))\n",
        "        mutation_indices = np.random.choice(copy_population.shape[0], num_mutations, replace=False)\n",
        "        mutation_pool = population[mutation_indices]\n",
        "        mutation_probabilities = [0.3, 0.1, 0.5, 0.1] # probabilities for selction, insert, inversion mostly\n",
        "        for chromosome in mutation_pool:\n",
        "            mutation = np.random.choice(self.MUTATION_TYPE) #, p=mutation_probabilities)\n",
        "            if mutation == 'insert_mutation':\n",
        "                mutant = self.insert_mutation(chromosome)\n",
        "            elif mutation == 'swap_mutation':\n",
        "                mutant = self.swap_mutation(chromosome)\n",
        "            elif mutation == 'scramble_mutation':\n",
        "                mutant = self.swap_mutation(chromosome)\n",
        "            else:\n",
        "                mutant = self.inversion_mutation(chromosome)\n",
        "            mutated_population.append(mutant)\n",
        "        return mutated_population\n",
        "\n",
        "    # Inserting allele somewhere\n",
        "    def insert_mutation(self, chromosome:np.array):\n",
        "        gen1, gen2 = np.random.choice(np.arange(len(chromosome)), 2, replace=False)\n",
        "        mask = chromosome != gen2\n",
        "        new_chromosome = chromosome[mask]\n",
        "        index_gen1 = np.where(new_chromosome == gen1)[0][0]\n",
        "        new_chromosome = np.insert(new_chromosome, index_gen1, gen2)\n",
        "        return new_chromosome\n",
        "\n",
        "    # Swapping two alleles\n",
        "    def swap_mutation(self, chromosome:np.array):\n",
        "        idx_1, idx_2 = np.random.choice(np.arange(len(chromosome)), 2, replace=False)\n",
        "        new_chromosome = np.copy(chromosome)\n",
        "        new_chromosome[idx_1] = chromosome[idx_2]\n",
        "        new_chromosome[idx_2] = chromosome[idx_1]\n",
        "        return new_chromosome\n",
        "\n",
        "    # Inverting some alleles\n",
        "    def inversion_mutation(self, chromosome:np.array):\n",
        "        idx_1, idx_2 = np.random.choice(np.arange(len(chromosome)), 2, replace=False)\n",
        "        new_chromosome = np.copy(chromosome)\n",
        "        split1 = np.min((idx_1, idx_2))\n",
        "        split2 = np.max((idx_1, idx_2))\n",
        "        reversed_sequence = chromosome[split1:split2+1][::-1] # +1 to ensure a minimum sequence of 2\n",
        "        new_chromosome[split1:split2+1] = reversed_sequence\n",
        "        return new_chromosome\n",
        "\n",
        "    # Scramble some alleles\n",
        "    def scramble_mutation(self, chromosome: np.array):\n",
        "        idx_1, idx_2 = np.random.choice(np.arange(len(chromosome)), 2, replace=False)\n",
        "        start, end = min(idx_1, idx_2), max(idx_1, idx_2)\n",
        "        # Scramble the segment\n",
        "        scrambled_segment = np.random.permutation(chromosome[start:end+1])\n",
        "        new_chromosome = np.copy(chromosome)\n",
        "        new_chromosome[start:end+1] = scrambled_segment\n",
        "        return new_chromosome\n",
        "\n",
        "    # Resetting population before starting new experiment\n",
        "    def reset(self, population_size=30):\n",
        "        self.POPULATION = np.array([np.random.permutation(self.tsp.dim) for _ in range(population_size)])\n",
        "\n",
        "    # Executing the GA according to the parameters\n",
        "    def experiment(self, N_REPETITIONS, N_TIMESTEPS, population=100, adaptive=True, mutation_rate=0.1, mutation='all',population_size=30):\n",
        "        repetition_array = np.zeros(N_TIMESTEPS)\n",
        "        repetition_array_b = np.zeros(N_TIMESTEPS)\n",
        "        for n in range(N_REPETITIONS):\n",
        "            self.reset(population_size)\n",
        "            timestep_array = np.zeros(N_TIMESTEPS)\n",
        "            timestep_array_b = np.zeros(N_TIMESTEPS)\n",
        "            population = self.POPULATION\n",
        "            fitness_scores = self.evaluate(population)\n",
        "            for t in range(N_TIMESTEPS):\n",
        "                timestep_array[t] =  np.mean(fitness_scores)\n",
        "                timestep_array_b[t] = np.min(fitness_scores)\n",
        "                parents = self.selection(fitness_scores, population,  Q=2, REPETITIONS=self.POPULATION_SIZE//2) #,  Q=2, REPETITIONS=self.POPULATION_SIZE/2)\n",
        "                offspring = self.cross_over(parents)\n",
        "                if adaptive==True:\n",
        "                  mutated_population = self.adaptive_mutation(offspring)\n",
        "                else:\n",
        "                  mutated_population = self.mutations(offspring, mutation_rate, mutation)\n",
        "                population = np.concatenate((parents, offspring, mutated_population))\n",
        "                fitness_scores = self.evaluate(population)\n",
        "            repetition_array += timestep_array / N_REPETITIONS\n",
        "            repetition_array_b += timestep_array_b / N_REPETITIONS\n",
        "            print(np.min(fitness_scores))\n",
        "            self.tsp.plot_route(population[np.argmin(fitness_scores)], np.min(fitness_scores))\n",
        "        return repetition_array, repetition_array_b\n",
        "\n",
        "    # Plotting the average and best path of genes\n",
        "    def plot(self, array, best_array):\n",
        "        plt.plot(array, color = 'skyblue', label = 'average route per population')\n",
        "        plt.plot(best_array, color = 'red', label = 'shortest route per population')\n",
        "        plt.xlabel('Timesteps')\n",
        "        plt.ylabel('Route Length in km')\n",
        "        plt.title('Route Lengths per Timestep')\n",
        "        plt.grid(True)\n",
        "        plt.legend()\n",
        "        plt.show()\n",
        "\n",
        "    # Experiment 1: evaluating mutation types\n",
        "    def mutation_comparison(self, N_REPETITIONS, N_TIMESTEPS, mutation_rate=0.1):\n",
        "      for mutation_t in self.MUTATION_TYPE:\n",
        "        array, _ = self.experiment(N_REPETITIONS, N_TIMESTEPS, False, mutation_rate=0.1, mutation=mutation_t)\n",
        "        plt.plot(array, linestyle = \"--\",label = mutation_t)\n",
        "      array, _ = self.experiment(N_REPETITIONS, N_TIMESTEPS, False, mutation_rate=0.1, mutation='all')\n",
        "      plt.plot(array, color = 'red', label = 'all mutations at random')\n",
        "      plt.xlabel('Timesteps')\n",
        "      plt.ylabel('Route Length in km')\n",
        "      plt.title('Mutation type comparison')\n",
        "      plt.grid(True)\n",
        "      plt.legend()\n",
        "      plt.show()\n",
        "\n",
        "    # Experiment 2: evaluating different mutation rates\n",
        "    def mutation_rate_comparison(self, N_REPETITIONS, N_TIMESTEPS, mutation_rate=0.1):\n",
        "      rates = np.linspace(0.05, 0.3, 10)\n",
        "      for i in rates:\n",
        "        array1, _ = self.experiment(N_REPETITIONS, N_TIMESTEPS, False, mutation_rate=i, mutation='all')\n",
        "        plt.plot(array1, linestyle='--', label = f'rate {i}')\n",
        "      plt.xlabel('Timesteps')\n",
        "      plt.ylabel('Route Length in km')\n",
        "      plt.title('Mutation_rate comparison')\n",
        "      plt.grid(True)\n",
        "      plt.legend(loc='upper left', bbox_to_anchor=(1, 1))\n",
        "      plt.show()\n",
        "\n",
        "    # Experiment 2: evaluating different population sizes\n",
        "    def population_comparison(self, N_REPETITIONS, N_TIMESTEPS, mutation_rate=0.1):\n",
        "      pops = np.linspace(10, 100, 10)\n",
        "      for i in pops:\n",
        "        array1, _ = self.experiment(N_REPETITIONS, N_TIMESTEPS, i, False, mutation_rate=0.1, mutation='all', population_size=i)\n",
        "        plt.plot(array1, linestyle='--', label = f'rate {i}')\n",
        "      plt.xlabel('Timesteps')\n",
        "      plt.ylabel('Route Length in km')\n",
        "      plt.title('Mutation_rate comparison')\n",
        "      plt.grid(True)\n",
        "      plt.legend(loc='upper left', bbox_to_anchor=(1, 1))\n",
        "      plt.show()\n",
        "\n",
        "    # Experiment 3: comparing adaptive to static GA\n",
        "    def adaptive_mutation_comparison(self, N_REPETITIONS, N_TIMESTEPS):\n",
        "      array1, _ = self.experiment(N_REPETITIONS, N_TIMESTEPS, False, mutation_rate=0.1, mutation='all')\n",
        "      array2, _ = self.experiment(N_REPETITIONS, N_TIMESTEPS, True, mutation_rate=0.1, mutation='all')\n",
        "      plt.plot(array1, color = 'skyblue', label = 'static')\n",
        "      plt.plot(array2, color = 'red', label = 'adaptive')\n",
        "      plt.xlabel('Timesteps')\n",
        "      plt.ylabel('Route Length in km')\n",
        "      plt.title('Adaptive vs Static mutation_rate')\n",
        "      plt.grid(True)\n",
        "      plt.legend()\n",
        "      plt.show()\n",
        "\n",
        "\n",
        "\n",
        "class RandomSearch():\n",
        "    \"\"\"Random searc, and plot of random path lenghts\"\"\"\n",
        "\n",
        "    def __init__(self, tsp: TSP, guesses):\n",
        "        self.guesses = guesses\n",
        "        self.tsp = tsp\n",
        "        self.best_path = None\n",
        "        self.best_length = float('inf')\n",
        "        self.best_length_t = np.zeros(guesses)\n",
        "        self.lengths = []\n",
        "\n",
        "    # Baseline: Evaluate random paths\n",
        "    def guessing_paths(self):\n",
        "        for i in range(self.guesses):\n",
        "            current_path = np.random.permutation(self.tsp.dim)\n",
        "            current_length = self.tsp(current_path)\n",
        "            self.lengths.append(current_length)\n",
        "            if current_length < self.best_length:\n",
        "                self.best_length =current_length\n",
        "                self.best_path = current_path\n",
        "            self.best_length_t[i]= self.best_length\n",
        "\n",
        "        return self.best_path, self.best_length, self.lengths, self.best_length_t\n",
        "\n",
        "    # Plot the fitness\n",
        "    def plot_random_paths(self):\n",
        "        plt.plot(self.best_length_t, color = 'skyblue', label = 'current shortest route', )\n",
        "        plt.xlabel('Timestep')\n",
        "        plt.ylabel('Route length in km')\n",
        "        plt.title('Average Route Length per Timestep')\n",
        "        plt.legend()\n",
        "        plt.grid(True)\n",
        "        # histogram\n",
        "        counts, bins = np.histogram(self.lengths, bins=30, density=True)\n",
        "        bin_centers = (bins[:-1] + bins[1:]) / 2\n",
        "        plt.figure(figsize=(10, 6))\n",
        "        plt.bar(bin_centers, counts, width=(bins[1] - bins[0]), alpha=0.7, color='skyblue', edgecolor='black')\n",
        "        # show best random path\n",
        "        plt.axvline(self.best_length, color='red', linestyle='--', linewidth=2, label=f'Shortest path found: {self.best_length:.2f}')\n",
        "        plt.axvline(np.mean(self.lengths), color='green', linestyle='--', linewidth=2, label=f'Average path length: {np.mean(self.lengths):.2f}')\n",
        "        # Label and Title\n",
        "        plt.xlabel('Path Length')\n",
        "        plt.ylabel('Percentage of occurence (%) ')\n",
        "        plt.title('Distribution of randomly generated path lengths')\n",
        "        plt.legend()\n",
        "        plt.show()\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    # Seed for reproducibility\n",
        "    np.random.seed(42)\n",
        "\n",
        "    # Experiment settings\n",
        "    POPULATION_SIZE = 30\n",
        "    N_REPETITIONS = 5\n",
        "    N_TIMESTEPS = 1000\n",
        "    GUESSES = 100000\n",
        "\n",
        "    # with TSP(plot=False) as tsp:\n",
        "    #     # Sample a 10 random paths\n",
        "    #     for _ in range(10):\n",
        "    #         # Sample a random path\n",
        "    #         random_path = np.random.permutation(tsp.dim)\n",
        "    #         # Here we can use tsp as a function\n",
        "    #         path_length = tsp(random_path)\n",
        "    #         # Display the path\n",
        "    #         tsp.plot_route(random_path, path_length)\n",
        "    #         # Input something to continue\n",
        "    #         input()\n",
        "\n",
        "    # Alternatively, you can use the object only for the TSP function,\n",
        "    # and dont use the plot\n",
        "    # with TSP(plot=True) as tsp:\n",
        "\n",
        "    # # Random baseline\n",
        "    # random_path = np.random.permutation(tsp.dim)\n",
        "    # route_length = tsp(random_path)\n",
        "    # print(f\"Random path length {route_length: .2f}km\")\n",
        "\n",
        "\n",
        "    # Plot best random paths search\n",
        "    # tsp = TSP(plot=False)\n",
        "    # ra = RandomSearch(tsp, guesses=GUESSES)\n",
        "    # ra.guessing_paths()\n",
        "    # ra.plot_random_paths()\n",
        "\n",
        "    # Initializing classes for GA\n",
        "    tsp = TSP(plot=False)\n",
        "\n",
        "    ga = GA(tsp, POPULATION_SIZE=POPULATION_SIZE)\n",
        "\n",
        "\n",
        "\n",
        "    # Testing: run a random experiment\n",
        "\n",
        "    #array, best_array = ga.experiment(N_REPETITIONS, N_TIMESTEPS, True, mutation_rate=0.1, mutation='all')\n",
        "    #ga.plot(array, best_array) #plot best path and average path over timesteps over repetitions\n",
        "\n",
        "    # Experiment 1: comparing mutations\n",
        "    #ga.mutation_comparison(N_REPETITIONS, N_TIMESTEPS, mutation_rate=0.1)\n",
        "\n",
        "    # Experiment 2: comparing hyperparameter settings for best mutations\n",
        "    ga.mutation_rate_comparison(N_REPETITIONS, N_TIMESTEPS, mutation_rate=0.1)\n",
        "    #ga.population_comparison(N_REPETITIONS, N_TIMESTEPS, mutation_rate=0.1) #does not work yet, check how population parameter gets passed on in experiment\n",
        "    # Experiment 3: comparing adaptive mutation GA to normal mutation GA\n",
        "    #ga.adaptive_mutation_comparison(N_REPETITIONS, N_TIMESTEPS)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "TD:\n",
        " - Experiment in Latex for comparison decription\n",
        " - probs for ratio. Improve hybrid GA.\n",
        " - Test parameters for best version of hybrid GA (mutation, population, Q maybe)\n",
        " - Experiment comparison for hybrid GA and hybrid adaptive GA\n",
        " - Check for random search implementation graphics reproducible\n",
        " - rerun mmutation rate comp with 0.01 to 0.8\n",
        " - do the plot again for single mutation and hybrid (with and without prob setting)\n",
        " - do the population optimization"
      ],
      "metadata": {
        "id": "6lhBztx708hh"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "2nd shortest distance found so far: 27701, 1000 timesteps, 3 reps ,seed 42, pop 30\n",
        "shortest distance: 25115., 2000 timesteps, 2 reps, seed 42, pop 30\n",
        "\n",
        "1st for adaptive: 24853, 2000 timesteps, 2 reps, seed 42, pop 30,\n",
        "\n",
        "\n",
        "1st for adaptive: 24842.5, 2000 timesteps, 3 reps, seed 42, pop 50, adapt comp with static,\n",
        "# New section\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "8kFX7M2T0pkX"
      }
    }
  ]
}